RDBC-992 - Introduce GenAI ETL Tasks, Configuration, and Operations

M4xymm opened the pull request:
**GenAI ETL Task Support**

- Introduced `GenAiConfiguration`, a detailed configuration class for GenAI ETL tasks, supporting validation, serialization, and management of AI-specific settings such as prompt templates, transformation scripts, concurrency, and response schemas.
- Added `GenAiTransformation` class to encapsulate and validate transformation scripts required for GenAI ETL, ensuring the presence of the necessary `ai.genContext` call.

**AI Task Identifier Management**

- Implemented `AiTaskIdentifierHelper` for generating and validating normalized, safe identifiers for AI tasks, enforcing naming conventions and providing normalization utilities.

**Operations for GenAI ETL Tasks**

- Added `AddGenAiOperation` and `UpdateGenAiOperation` classes (and their associated result types) to enable adding and updating GenAI ETL tasks on the server, including support for starting points and optional transformation state resets. [[1]](diffhunk://#diff-9ca0b0fa07240872354b0e1a76fab1e429b18b43eff7b089cc287b649b76cc23R1-R102) [[2]](diffhunk://#diff-e12ad496e99b77a82828dd77abcfe75c1950470ff769c58647127027474ee07bR1-R134) [[3]](diffhunk://#diff-5a408833763f6ce684d6dfb63a772801b7e9aea833a925ebe2c369822a49e5dfR1-R27)

**Base Class and Type Improvements**

- Introduced `AbstractAiIntegrationConfiguration` as a base class for AI integration configurations, extending ETL configuration with AI-specific properties and connection handling.
- Minor improvement to the `compare` method in `AiConnectionString` to simplify its signature.

M4xymm committed:
RDBC-992 feat: integrate GenAI configuration and operations into ETL framework

- Added `GenAiConfiguration` and `GenAiTransformation` for handling AI-related ETL tasks.
- Introduced `AddGenAiOperation` and `AddGenAiOperationResult` for new GenAI ETL task creation.
- Created `AiConnectionString` for managing AI service integrations (e.g., OpenAI, Azure, Hugging Face).
- Enhanced validation and identifier generation for AI task configurations.
- Updated ETL framework with `GenAi` and `EmbeddingsGeneration` types.

M4xymm committed:
RDBC-992 feat: add unit tests for GenAI ETL task operations

- Added tests for creating, updating, and validating GenAI ETL tasks.
- Introduced coverage for `AddGenAiOperation` and `UpdateGenAiOperation`.
- Validated error handling for missing schema or sample object in configurations.

RDBC-992 update test script validation logic, eslint fixes

kalczur commented on src/Documents/StartingPointChangeVector.ts:
```diff
@@ -0,0 +1,44 @@
+    private constructor(public value: StartingPointChangeVectorType) {
+    }
+
+    /**
+     * Creates a StartingPointChangeVector from a specific change vector string.
+     * @param changeVector The change vector string
+     * @returns A new StartingPointChangeVector instance
+     */
```
Many comments seem unnecessary. Is this a copy from c#?

M4xymm replied:
Yes, I left it because it is a 1:1 copy from C#

kalczur commented on test/Documents/Operations/AI/AiTaskIdentifierHelperTest.ts:
```diff
@@ -0,0 +1,158 @@
+
+describe("AiTaskIdentifierHelper", () => {
+    describe("validateIdentifier", () => {
+        it("should accept valid identifiers", () => {
+            const validIdentifiers = [
```
Should be in `it.each` instead of for loop?

M4xymm replied:
We use Mocha, which does not have `it.each`.

kalczur commented on test/Documents/Operations/AI/AiTaskIdentifierHelperTest.ts:
```diff
@@ -0,0 +1,158 @@
+                { id: "task/name" }
+            ];
+
+            for (const { id } of invalidIdentifiers) {
```
why now it's an object array?

kalczur commented on test/Documents/Operations/AI/GenAiConfigurationTest.ts:
```diff
@@ -0,0 +1,295 @@
+        it("should validate complete configuration successfully", () => {
+            const config = createValidConfiguration();
+
+            const errors = config.validate(false, false, false);
```
Mabye `validate` argument should be options object for readability?

kalczur commented on test/Documents/Operations/AI/GenAiConfigurationTest.ts:
```diff
@@ -0,0 +1,295 @@
+            const config = new GenAiConfiguration();
+            assert.strictEqual(config.etlType, "GenAi");
+        });
+
+        it("should return identifier as destination", () => {
```
IDK how it looks for other tasks but do we want to test such things like default values?

M4xymm replied:
I removed a few unnecessary tests

kalczur commented on test/Documents/Operations/AI/GenAiOperationsTest.ts:
```diff
@@ -0,0 +1,335 @@
+                StartingPointChangeVector.BeginningOfTime
+            );
+
+            assert.ok(operation);
```
add assert change vector

M4xymm replied:
After verification, the genAiOperationsTest file is not needed, due to GenAiEtlTest which performs similar tests.

kalczur commented on test/Documents/Operations/AI/GenAiOperationsTest.ts:
```diff
@@ -0,0 +1,335 @@
+            const conventions = new DocumentConventions();
+            const command = operation.getCommand(conventions);
+
+            assert.ok(command);
```
same

kalczur commented on test/Documents/Operations/AI/GenAiOperationsTest.ts:
```diff
@@ -0,0 +1,335 @@
+            const conventions = new DocumentConventions();
+            const command = operation.getCommand(conventions);
+
+            assert.ok(command);
```
Please check all test asserts

M4xymm committed:
RDBC-992 removed outdated GenAI operation tests and migrated validation to option-based API

- Deleted legacy `GenAiOperationsTest` and consolidated validation tests.
- Simplified `GenAiConfiguration.validate` to use options object for enhanced readability and flexibility.
- Refactored identifier validation tests for better coverage.
